### PR TITLE
Add the `autotools`, `configure` and `make` templates.

### PR DESCRIPTION
**Pull request by Arignir**

### CODE REVIEW

File: stdlib/template/__init__.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+"""Templates are functions that provide a default, configurable and extendable way to
+build packages of a certain type.
+
+There are different kinds of template:
+
+  * **Partial templates**: templates that provide only **a chunk** of the overall build process, like a
+    template that wraps a specific build system but don't download any source code or generate
+    any package.
+
+    They are usualy used as a sub-part of an *exhaustive template*.
```
Comment (Pedrochu) [inline stdlib/template/__init__.py]:
```suggestion
    They are usually used as a sub-part of an *exhaustive template*.
```

File: stdlib/template/__init__.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    Example of partial templates:
```
Comment (Pedrochu) [inline stdlib/template/__init__.py]:
```suggestion
    Examples of partial templates:
```

File: stdlib/template/autotools.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    extract=stdlib.extract.flat_extract_all,
+    patch=stdlib.patch.patch_all,
+    configure=configure,
+    compile=make,
+    check=lambda: make('check', fail_ok=True),
+    install=lambda: make('install', f'DESTDIR={stdlib.build.current_build().install_cache}'),
+    split=None,
+    deplinker=None,
+):
+    """Downloads, build and wraps a library based on ``autoconf`` and ``make``.
```
Comment (Pedrochu) [inline stdlib/template/autotools.py]:
```suggestion
    """Downloads, builds and wraps a library based on ``autoconf`` and ``make``.
```

File: stdlib/template/autotools.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    **Extract**
+
+        This step is used to extract the downloaded source code. The default value is :py:func:`.flat_extract_all` with no argument.
+
+    **Patch**
+
+        This step is used to patch the downloaded source code. The default value is :py:func:`.patch_all` with no argument.
+
+    From now one, the current working directory is changed in favor of ``build_folder`` (which defaults to ``.``).
```
Comment (Pedrochu) [inline stdlib/template/autotools.py]: If you meant to say 'starting from now', it should be written that way:
```suggestion
    From now on, the current working directory is changed in favor of ``build_folder`` (which defaults to ``.``).
```

File: stdlib/template/autotools.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        extract()
+
+    stdlib.log.ilog("Step 3/9: Patch")
+    if patch is not None:
+        patch()
+
+    os.makedirs(build_folder, exist_ok=True)
+    with stdlib.pushd(build_folder):
+
+        print(f"OMG I'M IN {os.getcwd()}")
```
Comment (Pedrochu) [inline stdlib/template/autotools.py]: is this some debug message you forgot to remove? :')

File: stdlib/template/configure.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+# -*- coding: utf-8 -*-
+"""Provides a small, partial template that wraps the ``configure`` command."""
+
+import os
+import stdlib
+from typing import List
+
+
+def get_dir_flags() -> List[str]:
+    """Build a list of directory flags used by the :py:func:`.configure` template.
```
Comment (Pedrochu) [inline stdlib/template/configure.py:11]:
```suggestion
    """Builds a list of directory flags used by the :py:func:`.configure` template.
```

File: stdlib/template/configure.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        '--datadir="/usr/share"',
+        '--infodir="/usr/share/info"',
+        '--mandir="/usr/share/man"',
+        '--sysconfdir="/etc"',
+        '--localstatedir="/var"',
+    ]
+
+
+def get_feature_flags() -> List[str]:
+    """Build a list of feature flags used by the :py:func:`.configure` template.
```
Comment (Pedrochu) [inline stdlib/template/configure.py:42]:
```suggestion
    """Builds a list of feature flags used by the :py:func:`.configure` template.
```

File: stdlib/template/configure.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+    The return value is a list of feature flags that enable parts of the target software based
```
Comment (Pedrochu) [inline stdlib/template/configure.py:44]:
```suggestion
    The return value is a list of feature flags that enables parts of the target software based
```

File: stdlib/template/configure.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        '--with-ncursesw',
+        '--with-pthreads',
+        '--with-threads',
+        '--with-readline',
+        '--with-history',
+    ]
+
+
+def get_flags() -> List[str]:
+    """Build a list of system-wide flags used by the :py:func:`.configure` template.
```
Comment (Pedrochu) [inline stdlib/template/configure.py:69]:
```suggestion
    """Builds a list of system-wide flags used by the :py:func:`.configure` template.
```

File: stdlib/template/configure.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    return get_dir_flags() + get_feature_flags()
+
+
+def configure(
+    *flags: str,
+    system_flags: bool = True,
+    binary: str = './configure',
+):
+    """Run ``./configure`` with a specific set of arguments.
```
Comment (Pedrochu) [inline stdlib/template/configure.py:81]:
```suggestion
    """Runs ``./configure`` with a specific set of arguments.
```

File: stdlib/template/configure.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+    The flags given to configure are both the return value of :py:func:`.get_flags()` and the given ``flags``.
```
Comment (Pedrochu) [inline stdlib/template/configure.py]: This sentence sounds a bit weird. Maybe you meant this?
```suggestion
    The flags given to configure are respectively the return value of :py:func:`.get_flags()` and the given ``flags``.
```

File: stdlib/template/configure.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+    Some configure script abort on unknown flags. If ``system_flags`` is ``False``, the default system flags
```
Comment (Pedrochu) [inline stdlib/template/configure.py]:
```suggestion
    Some configure script aborts on unknown flags. If ``system_flags`` is ``False``, the default system flags
```
Also did you meant to say `Some configure script`, and not something like `this configure script` or `the configure script`?

File: stdlib/template/configure.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+    :note: The prefix used is ``/usr`` (as returned by :py:func:`.get_dir_flags`). Therefore, the environment variable
+        ``DESTDIR`` **must** be set (pointing to the install cache of the current build) when installing the built software.
+        Otherwise, the installation can damage and overwrite parts of the host system.
+
+    :note: Because ``flags`` is appended at the end of the argument list, it overwrites any system-wide flags. This is useful
+        to easily disable a feature or package  with its ``--without-PACKAGE``/``--disable-FEATURE`` equivalent.
+
+    :param flags: A list of flags to give to the configure script.
+    :param system_flags: Indicate whether or not the template should include the system flags returned by :py:func:`.get_flags`.
```
Comment (Pedrochu) [inline stdlib/template/configure.py:104]:
```suggestion
    :param system_flags: Indicates whether or not the template should include the system flags returned by :py:func:`.get_flags`.
```

**Review by Pedrochu — DISMISSED** [review]

File: stdlib/template/__init__.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3.6
+# -*- coding: utf-8 -*-
+"""Templates are functions that provide a default, configurable and extendable way to
+build packages of a certain type.
+
+There are different kinds of template:
+
+  * **Partial templates**: templates that provide only **a chunk** of the overall build process, like a
+    template that wraps a specific build system but don't download any source code or generate
```
Comment (doom) [inline stdlib/template/__init__.py]:
```suggestion
    template that wraps a specific build system but doesn't download any source code or generate
```

File: stdlib/template/autotools.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        * ``configure``
+        * ``compile``
+        * ``check``
+        * ``install``
+        * ``split``
+        * ``dependency linking``
+
+    For each one of these steps, a function is called. This template simply calls all of them in the above order.
+    All of those functions can be given as arguments, but each one of them have a default value that is explained below.
+    If any of these function is ``None``, the step is skipped.
```
Comment (doom) [inline stdlib/template/autotools.py]:
```suggestion
    If any of those functions is ``None``, the step is skipped.
```

File: stdlib/template/autotools.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        * ``extract``
+        * ``patch``
+        * ``configure``
+        * ``compile``
+        * ``check``
+        * ``install``
+        * ``split``
+        * ``dependency linking``
+
+    For each one of these steps, a function is called. This template simply calls all of them in the above order.
```
Comment (doom) [inline stdlib/template/autotools.py]:
```suggestion
    For each one of these steps, a function is called. This template simply calls each of them in the above order.
```

File: stdlib/template/autotools.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    All of those functions can be given as arguments, but each one of them have a default value that is explained below.
```
Comment (doom) [inline stdlib/template/autotools.py]:
```suggestion
    All of these functions can be given as arguments, but each one of them has a default value that is explained below.
```

File: stdlib/template/autotools.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+
+        stdlib.log.ilog("Step 8/9: Split")
+        if split is not None:
+            split()
+
+        stdlib.log.ilog("Step 9/9: Dependency Linking")
+        if deplinker is not None:
+            deplinker()
+
+    return []
```
Comment (doom) [inline stdlib/template/autotools.py:137]: I think I need an explanation (and we probably need a doc comment) regarding why the empty list is returned here.

File: stdlib/template/configure.py (hunk 11 above)
Comment (doom) [reply stdlib/template/configure.py]: Also, you should annotate the "configure" with `py:func`

File: stdlib/template/configure.py (hunk 12 above)
Comment (doom) [inline stdlib/template/configure.py]:
```suggestion
    Some configure scripts abort on unknown flags. If ``system_flags`` is ``False``, the default system flags
```

Comment (doom) [reply stdlib/template/configure.py]: I think he meant to use a plural but forgot the "s" for "script"

File: stdlib/template/configure.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    The flags given to configure are both the return value of :py:func:`.get_flags()` and the given ``flags``.
+
+    Some configure script abort on unknown flags. If ``system_flags`` is ``False``, the default system flags
+    usually provided by this template won't be infused in the final call, making it more flexible for such
+    configure scripts.
+
+    In such cases, it is recommended to filter-out the bad flags from the return value of :py:func:`.get_flags` instead of
+    rebuilding them by hand and taking the risk of missing some of them.
+
+    Even if ``system_flags`` is ``False``, some core flags will remain infused.
```
Comment (doom) [inline stdlib/template/configure.py:92]: "infused" sounds weird to me here

File: stdlib/template/configure.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    If such flags are the cause of any errors from the configure script, it is recommended to run ``./configure``
```
Comment (doom) [inline stdlib/template/configure.py]:
```suggestion
    If such flags are the cause of any error from the configure script, it is recommended to run ``./configure``
```

File: stdlib/template/make.py [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+
+def make(
+    *targets: str,
+    binary: str = 'make',
+    folder: str = '.',
+    fail_ok: bool = False,
+):
+    """Run ``make``.
+
+    :note: Targets are all executed by one instance of ``make``, therefore they are run in parallel.
```
Comment (doom) [inline stdlib/template/make.py]:
```suggestion
    :note: Targets are each executed by one instance of ``make``, therefore they are run in parallel.
```

**Review by doom — CHANGES_REQUESTED** [review]

File: stdlib/template/autotools.py (hunk 5 above)
Comment (Arignir) [reply stdlib/template/autotools.py]: Naaaah, what makes you think that? :3

File: stdlib/template/configure.py (hunk 6 above)
Comment (Arignir) [reply stdlib/template/configure.py:11]: The current coding style for NBuild is to use infinitive verb to document a function (like commit messages)

File: stdlib/template/autotools.py (hunk 18 above)
Comment (Arignir) [reply stdlib/template/autotools.py:137]: That is temporary.

The return value should be the one of the splitter, but this part of Nbuild isn't revamped yet, so there is no return value so far.

File: stdlib/template/autotools.py [hunk 22]
```diff
@@ -26,7 +26,7 @@ def build(
     split=None,
     deplinker=None,
 ):
-    """Downloads, build and wraps a library based on ``autoconf`` and ``make``.
+    """Downloads, builds and wraps a library based on ``autoconf`` and ``make``.
```
Comment (doom) [inline stdlib/template/autotools.py]:
```suggestion
    """Download, build and wrap a library based on ``autoconf`` and ``make``.
```

**Review by doom — CHANGES_REQUESTED** [review]

**Review by doom — APPROVED** [review]

**Pull request merged**